### DISCUSSION TITLE
Migrating Vue / Spring Boot app to Authelia

### DISCUSSION THREAD

**Discussion by ginkel** [authelia/authelia#7415] [Q&A]
Hi there,

I'm currently in the process of migrating our applications from Keycloak to Authelia. One application that I could not yet migrate successfully is an in-house application that uses Vue for its frontend and Spring Boot for its backend.

The client config in Authelia looks like this:

```
  - client_id: *redacted*
    client_name: *redacted*
    client_secret: *redacted*
    public: false
    redirect_uris:
      - https://*redacted*/oidc/callback
    scopes: "{{ authelia_oidc_default_scopes }}"
    authorization_policy: one_factor
    consent_mode: implicit
    grant_types:
      - implicit
    response_types:
      - id_token token
    userinfo_signed_response_alg: none
```

The Vue seems to authenticate successfully and is using Bearer authentication for all backend calls, such as in:

```
Authorization:
Bearer authelia_at_PC7LdUEzonE0VOePcJPvcTD7PEShb4Nx8yUDeKgkJo4.POF9ROREWbE5hvzb-k8vMvEf8hLbe73Qo_KyPseWwhg
```

Unfortunately, the backend responds with a 401 and the following error:

```
Www-Authenticate:
Bearer error="invalid_token", error_description="An error occurred while attempting to decode the Jwt: Invalid unsecured/JWS/JWE header: Invalid JSON: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $", error_uri="https://tools.ietf.org/html/rfc6750#section-3.1"
```

From its logs:

```
2024-06-12T16:34:24.047Z DEBUG 1 --- [r-http-epoll-12] o.s.s.w.s.a.AuthenticationWebFilter      : Authentication failed: An error occurred while attempting to decode the Jwt: Invalid unsecured/JWS/JWE header: Invalid JSON: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $

org.springframework.security.oauth2.server.resource.InvalidBearerTokenException: An error occurred while attempting to decode the Jwt: Invalid unsecured/JWS/JWE header: Invalid JSON: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
        at org.springframework.security.oauth2.server.resource.authentication.JwtReactiveAuthenticationManager.onError(JwtReactiveAuthenticationManager.java:79)
[...]
```

Unfortunately, I can't seem to find the malformed payload in the logs, so I'm somewhat clueless what it's seeing instead of the expected format.

The Spring Boot documentation is located at: https://docs.spring.io/spring-security/reference/servlet/oauth2/resource-server/jwt.html

Any ideas how to figure out what's wrong are much appreciated!

Thanks,
Thilo

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Hi Thilo,

It looks like the issue you're encountering is related to the JWT (JSON Web Token) format that Authelia is generating and how Spring Boot is attempting to decode it. The error message indicates that Spring Boot is expecting a JSON object but is receiving a string instead.

Here are a few steps you can take to troubleshoot and potentially resolve this issue:

1. **Check the JWT Format**:
   Ensure that the JWT generated by Authelia is in the correct format. A JWT typically consists of three parts: Header, Payload, and Signature, separated by dots (.). It should look something like this:
   ```
   eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c
   ```

2. **Verify the Token**:
   Use a tool like [jwt.io](https://jwt.io/) to decode the token and verify its structure. This can help you see if the token is malformed or if there are any issues with the header or payload.

3. **Authelia Configuration**:
   Ensure that your Authelia configuration is correct and that it is generating a valid JWT. Based on your provided configuration, it looks like you are using the `implicit` grant type and `id_token token` response type. Make sure these settings are appropriate for your use case.

4. **Spring Boot Configuration**:
   Double-check your Spring Boot configuration to ensure it is set up to handle JWTs correctly. Refer to the [Spring Security documentation](https://docs.spring.io/spring-security/reference/servlet/oauth2/resource-server/jwt.html) for guidance on configuring JWT decoding.

5. **Enable Debug Logging**:
   Enable debug logging in both Authelia and Spring Boot to get more detailed logs. This can help you identify where the issue might be occurring.

6. **Check for Known Issues**:
   Look for any known issues or discussions related to JWT decoding in Authelia and Spring Boot. For example, you can check the [Authelia GitHub issues](https://github.com/authelia/authelia/issues) and [discussions](https://github.com/authelia/authelia/discussions) for similar problems.

If you continue to face issues, please provide more details about the JWT token structure and any additional logs that might help in diagnosing the problem.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by ginkel** [comment]
If I understand this correctly, Spring Boot (at least how the application is currently built) is expecting a JWT for stateless validation. Authelia probably submits an opaque token.

I changed this by setting `access_token_signed_response_alg: RS256`, now I got a different error to analyze ;-)

```
Bearer error="invalid_token", error_description="Failed to validate the token", error_uri="https://tools.ietf.org/html/rfc6750#section-3.1"
```

**Comment by ginkel** [comment]
Some debug logging later, this seems to be the current culprit: `BadJOSEException: JOSE header typ (type) at+jwt not allowed`

**Comment by james-d-elliott** [comment]
The access token is used to access resources and there are no guarantees or contracts between providers and relying parties as to the format of the access token. The format of this token is effectively opaque under all circumstances and should be treated as such.

It can be used to access resources secured by the provider or resource servers configured to introspect the token with the provider. If the access token is configured to be a JWT we follow the format of the standard: https://datatracker.ietf.org/doc/html/rfc9068#section-2.1

The ID token is used to prove authentication, and has an agreed contract on the format. I'd guess that's what it's actually expecting.

**Comment by ginkel** [comment]
I got it working by switching the application to support opaque tokens. Thanks for your help!
